### DISCUSSION TITLE
How to update SourceCache when its item's property is changed

### DISCUSSION THREAD

**Discussion by cocoderus** [reactivemarbles/DynamicData#952] [Q&A]
I've assigned a source cache to the data grid, and I need to save any updates made to the item. Each item has a command that changes a property, and this command is connected to a button in the data grid cell. The property is updating, and it's visible in the ui, but when I filter the source cache and then reset the filter, the property is also reset. How can I save the changes?

**Comment by Snailya** [comment]
Hello,
Are you assigning the source cache to data grid by calling bind method or directly binding in the UI?
```
cache.Connect()
.Bind(out _data)
.Subscribe()
```
If you are using the out data, the change is not happen in source cache. You need to update the source cache by calling cache.AddOrUpdate(items)

**Comment by JakenVeina** [comment]
Gonna need to see what your query looks like.

From the sound of it, I'd guess that you have a setup where resetting the filter triggers a transform to re-evaluate for all items, thus effectively resetting them all to an original state that you're not updating.

**Reply by cocoderus** [reply]
My query looks like this 

```c#
_sourceCache.Connect()
    .Filter(this.WhenAnyValue(a => a.SearchQuery)
        .Throttle(TimeSpan.FromSeconds(1))
        .DistinctUntilChanged()
        .Select(Filter))
    .ObserveOn(RxApp.MainThreadScheduler)
    .Do(_ => IsLoading = true)
    .ObserveOn(RxApp.TaskpoolScheduler)
    .ObserveOn(RxApp.MainThreadScheduler)
    .TransformToTree(x => x.ParentOrganizationId)
    .Transform(node => new Organization(node))
    .DisposeMany()
    .SortAndBind(out _source, SortExpressionComparer<Organization>.Ascending(a => a.OrganizationName))
    .Subscribe(_ => IsLoading = false);
```

**Reply by JakenVeina** [reply]
`Organization` here being your ViewModel, yes? What does your command logic look like?

**Reply by cocoderus** [reply]
Yes, Organization is vm which is ReactiveObject, and command is a simple method that changes property 
```c#
public void Update()
{
    IsFavorite = !IsFavorite;
}
```

**Reply by JakenVeina** [reply]
And I presume the setter for `IsFavorite` is not writing back to `_sourceCache`? So yeah, when a change IS made to `_sourceCache` that triggers the entire `Organization` object to get rebuilt and re-subscribed, which is highly inefficient, although not immediately helpful to know.

Ideally, you want to figure out what action is triggering `transformFactory` to re-run for the item, and maybe fix that. You can toss a `.Do()` in before `.Transform()` and inspect the changesets directly, to see.

You can also swap `.Transform()` to `.TransformWithInlineUpdate()` This will invoke an `updateAction` that you provide, whenever updates come in, instead of re-invoking `transformFactory` to create a new `Organization` instance each time. So, any transient state in the `Organization` that hasn't been saved back to `_sourceCache` will be preserved. Honestly, this is the operator that you should almost always be using for building VMs.

**Reply by cocoderus** [reply]
I've created a [sample project](https://github.com/cocoderus/DDTest) to show what I'm trying to achieve. I need to display large amounts of data in a tree-like structure and be able to filter and update certain properties. I noticed that in my current implementation, `.Transform()` uses a lot of memory when filtering or refreshing the data. Therefore, if possible, I would appreciate it if you could show me how to do this correctly in a sample project.

**Reply by JakenVeina** [reply]
Yes, I'd very much like to do that. Sounds like a great excuse for me to learn what `.TransformToTree()` is really capable of (or not capable of). I'm very unfamiliar with it, myself.

It might not be until this weekend that I can dedicate some time to it.

**Reply by JakenVeina** [reply]
@cocoderus Alright, I've been playing with this for a bit now, and my initial take is... this is a pain in the ass.

I'm gonna have to start by asking to refine the parameters a little more, because I noticed something about your given implementation that I"m guessing is unintentional: The whole tree concept is effectively broken when doing a search.

I'll explain.

Suppose we have the following tree:

```
Jake
 * Coco
   * Roland
   * Darrin
 * Isaac
   * Chris
Colette
```

Now, let's say we filter for "a". With the `.Filter()` being applied before `.TransformToTree()`, what we end up with in the UI is...

```
Jake
 * Isaac
Roland
Darrin
```

Sure, we can see all the items that matched the filter, no matter where they were in the tree, but the tree itself is almost completely lost. Because "Coco" was removed before rolling up the tree, we've lost the ancestral relationship between "Jake", "Roland", and "Darrin". They now appear as siblings at the top level of the tree, which really doesn't represent the reality of the data.

Now, maybe this isn't a problem in whatever your scenario is, but to me, there's a couple ways to make "filtering within a tree" make sense, and this isn't one of them. Using the same example tree...

1) Just filter at the top level
```
Jake
 * Coco
   * Roland
   * Darrin
 * Isaac
   * Chris
```

Top level items are "Jake" and "Colette" and we just keep the only one that has an "a". We don't bother applying the filtering at the lower levels of the tree. If you want that, you would do some kind of sub-search within the expansion.

2) Filter to include items that match, and also items with matching children, recursively.
```
Jake
 * Coco
   * Roland
   * Darrin
 * Isaac
```

This one for sure makes the most usability sense, to me. We see all the items that ACTUALLY match, same as before, but they retain their position within the tree. We end up including "Coco" because if we didn't, we wouldn't be able to see "Roland" and "Darrin".

Option 2 is what I was trying to implement as an alternative for you, but it's proven to be quite challenging. I've been able to get it to work with ~1000 items, but for 65k, it's entirely infeasible. I would need to think through a way to model and implement this without recursion, because the recursion at that scale is murder on performance.

Long story short, at the moment, `.TransformToTree()` is pretty cool and useful, but it really isn't compatible with filtering. If you want to see the code, let me know. Otherwise, I'd be interested in continuing experimentation, but I"m hoping for some more context for what you're trying to do.

**Reply by cocoderus** [reply]
Thank you for taking the time to look into this. I want to clarify that there is nothing unintentional about my implementation. The first result you provided is exactly what I want as a result of the filtering operation.

In my current scenario, it is not important where the item is in the tree. The purpose of filtering is to find all matching items in the source. The tree structure only matters if the item's child or parent item also matches the search query.

**Reply by JakenVeina** [reply]
> Therefore, if possible, I would appreciate it if you could show me how to do this correctly in a sample project.

I wouldn't call it "correct", more just how I would implement this problem. It's largely the same as what you're doing, with a little more emphasis on state management and post-back, and using the `.TransformWithInlineUpdate()` operator, which should translate to less memory churn to drive the UI.

https://github.com/JakenVeina/DynamicDataTreeDemo
